x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/jwt/JwtRealm.java

```diff
@@ -401,8 +406,14 @@ public void authenticate(final AuthenticationToken authenticationToken, final Ac
                 // Intercept the role mapper listener response to log the resolved roles here. Empty is OK.
                 final String[] rolesArray = rolesSet.toArray(new String[0]);
                 final User user = new User(principal, rolesArray, null, null, userData.getMetadata(), true);
-                final String rolesString = Arrays.toString(rolesArray);
-                LOGGER.debug("Realm [" + super.name() + "] mapped roles " + rolesString + " for principal=[" + principal + "].");
+                LOGGER.debug(
+                    () -> new ParameterizedMessage(
```

Comment:
Why lambda?

Reply:
Because building the comma separated string is not-free, so it should be deferred to only happen if debug is enabled.

Our typical idioms for low-cost `debug` or `trace` logging are, in order of preference
1. A static string. Cheap, but often low value since it's the same message every time.
2. A simple log with parameters (`"[{}]"`), as long as (a) the parameters are low cost to pass in (b) there's no exception. If there is an exception there's potential confusion between the arguments that are parameters and the argument for the exception
3. A simple `logger.debug(new ParameterizedMessage("Message here [{}]", arg), ex)`, when the parameters are low cost to pass in, but there is an exception. The parameters go inside the message, but the exception is outside the message so there is less confusion between them.
4. A lambda supplier of a  `ParameterizedMessage(...)` for when the parameters are not low-cost, and should only be evaluated if the log message actually going to be output
5. Wrapping in `if (logger.isDebugEnabled())`, which is typically more intrusive in the code, but can sometimes end up simpler/neater than a lambda if there are several parameters to evaluate,